### PR TITLE
Hot path optimization for ReadValue/WriteValueIntoState for most used controls

### PR DESCRIPTION
**Pull request by jimon**
### Description

Current performance of `ReadValue` leaves a lot to be desired.

This PR is based on observation that for majority of controls:
- Memory layout is typical for the type, e.g. `AxisControl` is byte aligned float, `QuaternionControl` is byte aligned `Quaternion`, etc.
- No extra processing is applied.

### Changes made

- If memory layout is expected typical layout for the control type, we short circuit `ReadValue` to cast state pointer directly to expected type, bypassing `InputStateBlock` reading all together for selected controls. Same for `WriteValueIntoState`.
- Added control debugger visibility to check if control got hot path optimized or not.
- Require `ApplyParameterChanges` for application of control changes that change optimization status, like `AxisControl` processors parameters (e.g. `invert`, etc).

### Notes

We plan to land this feature as opt-in for now. And make it opt-out after some rigorous testing.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs [hunk 1]
```diff
@@ -190,8 +192,18 @@ public class PoseControl : InputControl<PoseState>
         /// </remarks>
         public PoseControl()
         {
-            m_StateBlock.format = new FourCC('P', 'o', 's', 'e');
+            m_StateBlock.format = PoseState.s_Format;
         }
+        
+        private void TryEnableFastPath()
+        {
+            m_FastPathReading = m_StateBlock.format == PoseState.s_Format;
```
Comment (jimon) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs]: Ok, after some thinking I think this is not sufficient, we also need to go through every subcontrol and check if there are no subprocessors on axis controls, etc.

Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs]: Same applies for other composite types like vector3/quaternion/etc

Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs [hunk 2]
```diff
@@ -75,6 +76,7 @@ public ButtonControl()
         /// <returns>True if <paramref name="value"/> crosses the threshold to be considered pressed.</returns>
         /// <seealso cref="pressPoint"/>
         /// <seealso cref="InputSettings.defaultButtonPressPoint"/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs:79]: I'm a bit sceptical that all the aggressive inlining is going to actually speed anything up. Probably the opposite.

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 3]
```diff
@@ -853,6 +854,71 @@ protected internal uint stateOffsetRelativeToDeviceRoot
...
+
+            CalculateOptimizedControlDataType();
+        }
+
+        // This function exists to warn users to start using ApplyParameterChanges for edge cases that were previously not intentionally worked,
+        // where control properties suddenly change underneath us without us anticipating that.
+        // This is mainly to AxisControl fields being public and capable of changing at any time even if we don't thing it's an anticipated usage pattern.
+        // Also it's not clear if InputControl.stateBlock.format can potentially change at any time, likely not.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void EnsureOptimizationTypeDoesntSuddenlyChange()
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs]: Feel a more imperative-style name works better for this method. EnsureOptimizationTypeHasNotChanged

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 4]
```diff
@@ -853,6 +854,71 @@ protected internal uint stateOffsetRelativeToDeviceRoot
...
+                    inputControl.CalculateOptimizedControlDataTypeRecursively();
+
+            CalculateOptimizedControlDataType();
+        }
+
+        // This function exists to warn users to start using ApplyParameterChanges for edge cases that were previously not intentionally worked,
+        // where control properties suddenly change underneath us without us anticipating that.
+        // This is mainly to AxisControl fields being public and capable of changing at any time even if we don't thing it's an anticipated usage pattern.
+        // Also it's not clear if InputControl.stateBlock.format can potentially change at any time, likely not.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:936]: Maybe mark this as [Conditional("DEBUG")] instead of using #IF DEBUG, so there's zero cost at runtime. Just a method call but still.

File: Packages/com.unity.inputsystem/InputSystem/Controls/AxisControl.cs [hunk 5]
```diff
@@ -277,5 +304,28 @@ public override unsafe float EvaluateMagnitude(void* statePtr)
...
+                m_StateBlock.bitOffset == 0)
+                return OptimizedControlDataType.Float;
+            if (noProcessingNeeded &&
+                     m_StateBlock.format == InputStateBlock.FormatBit &&
+                     // has to be 8, otherwise we might be mapping to a state which only represents first bit in the byte, while other bits might map to some other controls
+                     // like in the mouse where LMB/RMB map to the same byte, just LMB maps to first bit and RMB maps to second bit
+                     m_StateBlock.sizeInBits == 8 &&
+                     m_StateBlock.bitOffset == 0)
+                return OptimizedControlDataType.BoolAsByte;
+            return OptimizedControlDataType.NotOptimized;
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Controls/AxisControl.cs]: I think we might also need to check if the control uses state from another control (usesStateFromOtherControl) and make sure the other control also has non-default processing if so, potentially following a chain of state from other controls. This is getting messy

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs [hunk 6]
```diff
@@ -209,26 +211,66 @@ protected override void FinishSetup()
...
+                    velocity.WriteValueIntoState(value.velocity, statePtr);
+                    angularVelocity.WriteValueIntoState(value.angularVelocity, statePtr);
+                    break;
+            }
+        }
+        
+        public OptimizedControlDataType CalculateOptimizedControlDataType()
+        {
+            if (
+                m_StateBlock.sizeInBits == 60 * 8 &&
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs]: Should be PoseState.kSizeInBytes instead of 60?

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
+﻿namespace UnityEngine.InputSystem.Controls
+{
+    public enum OptimizedControlDataType
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: Not sure about this design. Users wouldn't be able to optimize their own controls with this. Can it not be an actual Type type?

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        Quaternion,
+
+        // PoseState
+        Pose
+    }
+    
+    internal interface IOptimizedControl
+    {
+        public OptimizedControlDataType CalculateOptimizedControlDataType();
+    }
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: Could we get away with an isOptimized property on the base InputControl class instead of the interface?

**Review by andrew-oc — COMMENTED** [review]
Really cool direction. Have some concerns about the implementation and I'm also strongly opposed to ever making it an opt-out behavior in any 1.x version, but I like it so far.

File: Packages/com.unity.inputsystem/InputSystem/Controls/AxisControl.cs (hunk 5 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/AxisControl.cs]: I though if we use state from other control we just do that .. use the state memory. As I don't understand how or where would processor stack from another control run.

Is there an example where one control uses state from another control and when reading first control we run processor stack from second control?

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs (hunk 7 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: I want to easily switch case on the value, so enum would be better/faster.
It doesn't have to be public, we can make it internal. I just had struggle with internal interface public method, C# somehow wants this enum type being public, but we can wiggle around it I think.

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs (hunk 8 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: No, we need to cache the value somehow, because for example for `Vector3Control` to decide that it can be optimized it needs to know if x/y/z controls are optimized and are of float type

File: Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs (hunk 2 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs:79]: Why? These checks are in the hot path, doing it inline _should_ be faster. But we don't have a good indicative test suit sigh

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs (hunk 7 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: Not sure about users optimizing their own controls, this feature would be internal for a time being, if we get nudged we might consider opening it.

In general I'm not convinced that user defined types are a good idea, that greatly limits what we can do with regards to editor UX and data structures. If all controls types are known at our compile time we would be able to do much more effective architecture.

Users defining their own devices is different and should stay as-is.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs (hunk 6 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs]: Will fix but it doesn't matter that much, as you can see down in the check we also check layout of controls, so if anything about `PoseState` changes we also need to change that code as well.

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 4 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:936]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 3 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs (hunk 7 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: tried with Type and the checks are much more expensive it seems https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC2A3utmbjgKYB2ArlNgCqUTACidUp5Ja5/2AILYAvNgAMAGm4DsAIWl9ZBGdgC+AblWqADgCcAlgDcwwSkxbtO2M6w71RFu5wB0grUrL7jp84wCeOua2wAFBjsCBlAD2AGYAFAbUwACUHvyqWNgARtHRADbY9lD4ABaUAMYA1vEpxKr8cADsNpbFomLMzvQu+OnkauiZOLkFTFFllTV1vMq4LSFh5iJikUFx8bH50aZpqoNoakA===

Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: After some discussion, we can switch to int (FourCC) the same way as `InputStateBlock` works

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        // Quaternion as 4 consecutive 32 bit floats, in order of x,y,z,w, byte aligned
+        Quaternion,
+
+        // PoseState
+        Pose
+    }
+    
+    internal interface IOptimizedControl
+    {
+        public OptimizedControlDataType CalculateOptimizedControlDataType();
```
Comment (jimon) [inline Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: Ok that's not gonna fly:
- Let's say we have a custom control that derives from `AxisControl` and does custom processing in `ReadUnprocessedValueFromState`
- We make `Vector3Control` of such controls
- Now `Vector3Control` has no way to check if the nested controls are optimized or not, because interface is already implemented on `AxisControl` level and can't be implemented again in derived controls.

Suggestion: use virtual method on `InputControl` instead

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs (hunk 6 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs (hunk 7 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs (hunk 9 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs (hunk 8 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/IOptimizedControl.cs]: N/A, interface removed

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 10]
```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
...
+
+        /// <summary>
+        /// Apply built-in parameters changes (e.g. <see cref="AxisControl.invert"/>, others) and recompute <see cref="InputControl.optimizedControlDataType"/> for impacted controls.
+        /// </summary>
+        /// <remarks>
+        /// </remarks>
+        public void ApplyParameterChanges()
+        {
+            // First we go through all children of our own hierarchy
+            SetOptimizedControlDataTypeRecursively();
```
Comment (jimon) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:895]: Or just computing status for current control is sufficient?

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 11]
```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
...
+        }
+
+        // This function exists to warn users to start using ApplyParameterChanges for edge cases that were previously not intentionally supported,
+        // where control properties suddenly change underneath us without us anticipating that.
+        // This is mainly to AxisControl fields being public and capable of changing at any time even if we were not anticipated such a usage pattern.
+        // Also it's not clear if InputControl.stateBlock.format can potentially change at any time, likely not.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        // Only do this check in debug in hope that it will be sufficient to catch any misuse during development.
+        [Conditional("DEBUG")]
+        internal void EnsureOptimizationTypeHasNotChanged()
```
Comment (jamesmcgill) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:939]: Should this function go a bit further and ensure that `optimizeddControlDataType` is correct before continuing?

File: Assets/Tests/InputSystem/CoreTests_Controls.cs [hunk 12]
```diff
@@ -1378,4 +1378,45 @@ public void Controls_TouchControlStateCorrespondsToTouchState()
...
+
+        var mouse = InputSystem.AddDevice<Mouse>();
+
+        mouse.position.x.invert = true;
+        mouse.position.x.ApplyParameterChanges();
+
+        Assert.That(mouse.position.x.optimizedControlDataType, Is.EqualTo(InputStateBlock.FormatInvalid));
+        Assert.That(mouse.position.y.optimizedControlDataType, Is.EqualTo(InputStateBlock.FormatFloat));
+        Assert.That(mouse.position.optimizedControlDataType, Is.EqualTo(InputStateBlock.FormatInvalid));
+    }
```
Comment (jamesmcgill) [inline Assets/Tests/InputSystem/CoreTests_Controls.cs:1421]: Maybe a quick test that the values coming out are correct (especially for quaternion would be nice), or hijack some existing tests by making the feature flag state conditional on TestCase?

File: Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs (hunk 2 above)
Comment (jamesmcgill) [reply Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs:79]: Inclined to agree with Andrew, if we don't have good evidence otherwise, I would leave the compiler unconstrained to figure it out by itself.

**Review by jamesmcgill — APPROVED** [review]
Nice optimisation!

File: Assets/Tests/InputSystem/CoreTests_Controls.cs (hunk 12 above)
Comment (ekcoh) [reply Assets/Tests/InputSystem/CoreTests_Controls.cs:1421]: +1 to what @jamesmcgill  said. Definately sounds best to parameterize existing tests since they should be invariant to optimization setting if functional tests right?

Comment (ekcoh) [reply Assets/Tests/InputSystem/CoreTests_Controls.cs:1421]: In addition related to testing, would it be possible to cover this performance targeted feature with performance tests, as in "Automation/Performance Testing" in internaldocs and benefit from that? This relates to discussion of having monitoring of performance KPIs and I understand if out of scope, but interested in understand if this would be feasable. It would give reporting and monitoring of any changes in this and a diff without doing manual profiling?

File: Packages/com.unity.inputsystem/CHANGELOG.md [hunk 13]
```diff
@@ -16,6 +16,10 @@ however, it has to be formatted properly to pass verification tests.
 ### Actions
 - Extended input action code generator (`InputActionCodeGenerator.cs`) to support optional registration and unregistration of callbacks for multiple callback instances via `AddCallbacks(...)` and `RemoveCallbacks(...)` part of the generated code. Contribution by [Ramobo](https://github.com/Ramobo) in [#889](https://github.com/Unity-Technologies/InputSystem/pull/889).
 
+### Added
+
+- Significantly optimized cost of `ReadValue`/`ReadUnprocessedValueFromState`/`WriteValueIntoState` for some control types. Optimization is opt-in for now, please call `InputSystem.settings.SetInternalFeatureFlag("USE_OPTIMIZED_CONTROLS", true);` in your project to enable it. You can observe which controls are optimized by looking at new optimized column in device debugger. You might need to call a new `InputControl.ApplyParameterChanges()` method if the code is changing `AxisControl` fields after initial setup is done.
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: Regarding "You might need to call....", I assume this is required for the given condition and otherwise not? Maybe the correct approach is to only state that its required to be called for changes to have effect or what is the behavior if failing to do this?

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 14]
```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
...
+        {
+            return InputStateBlock.kFormatInvalid;
+        }
+
+        /// <summary>
+        /// Apply built-in parameters changes (e.g. <see cref="AxisControl.invert"/>, others) and recompute <see cref="InputControl.optimizedControlDataType"/> for impacted controls.
+        /// </summary>
+        /// <remarks>
+        /// </remarks>
+        public void ApplyParameterChanges()
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:892]: Looks like this is public API in public class right? Suggest to include @duckets on review when adding public API docs to get review feedback.

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 15]
```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
...
+            m_OptimizedControlDataType = InputSettings.optimizedControlsFeatureEnabled
+                ? CalculateOptimizedControlDataType()
+                : (FourCC)InputStateBlock.kFormatInvalid;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void SetOptimizedControlDataTypeRecursively()
+        {
+            // Need to go depth-first because CalculateOptimizedControlDataType might depend on computed values of children
+            if (m_ChildCount > 0)
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:923]: Would recommend adding braces to this if to avoid later mistakes on what is contained in it when doing multi-line statement

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs [hunk 16]
```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
...
+        // Also it's not clear if InputControl.stateBlock.format can potentially change at any time, likely not.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        // Only do this check in debug in hope that it will be sufficient to catch any misuse during development.
+        [Conditional("DEBUG")]
+        internal void EnsureOptimizationTypeHasNotChanged()
+        {
+            if (!InputSettings.optimizedControlsFeatureEnabled)
+                return;
+
+            if (CalculateOptimizedControlDataType() != optimizedControlDataType)
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs]: Good to see that error message contains info to track down the issue!

File: Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs [hunk 17]
```diff
@@ -66,17 +67,61 @@ protected override void FinishSetup()
...
+            switch (m_OptimizedControlDataType)
+            {
+                case InputStateBlock.kFormatQuaternion:
+                    return *(Quaternion*)((byte*)statePtr + (int)m_StateBlock.byteOffset);
+                default:
+                    return new Quaternion(
+                        x.ReadValueFromState(statePtr),
+                        y.ReadValueFromState(statePtr),
+                        z.ReadValueFromState(statePtr),
+                        w.ReadUnprocessedValueFromState(statePtr));
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs:79]: If there was a conclusion why we have this odd patern of w not being processed, maybe add a comment explaining/motivating it if information was fresh?

File: Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs [hunk 18]
```diff
@@ -66,17 +67,61 @@ protected override void FinishSetup()
...
+                    z.WriteValueIntoState(value.z, statePtr);
+                    w.WriteValueIntoState(value.w, statePtr);
+                    break;
+            }
+        }
+
+        protected override FourCC CalculateOptimizedControlDataType()
+        {
+            if (
+                m_StateBlock.sizeInBits == 128 &&
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs]: Why is this 128, get bit of magic number associations, would it be more clear with something like?
m_StateBlock.sizeInBits == sizeof(float) * 4 + sizeof(overhead....)
or similar?

File: Packages/com.unity.inputsystem/InputSystem/Controls/Vector3Control.cs [hunk 19]
```diff
@@ -31,23 +32,57 @@ protected override void FinishSetup()
...
         public override unsafe float EvaluateMagnitude(void* statePtr)
         {
             ////REVIEW: this can go beyond 1; that okay?
             return ReadValueFromState(statePtr).magnitude;
         }
+
+        protected override FourCC CalculateOptimizedControlDataType()
+        {
+            if (
+                m_StateBlock.sizeInBits == 96 &&
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/Vector3Control.cs]: Similar comment like on previous fixed size number

File: Packages/com.unity.inputsystem/InputSystem/Editor/Internal/InputControlTreeView.cs [hunk 20]
```diff
@@ -184,30 +189,18 @@ private ControlItem BuildControlTreeRecursive(InputControl control, int depth, r
...
-                offset = new GUIContent(offset.ToString()),
-                bit = new GUIContent(control.stateBlock.bitOffset.ToString()),
-                sizeInBits = new GUIContent(control.stateBlock.sizeInBits.ToString()),
-                type = new GUIContent(control.GetType().Name),
-                value = value,
-                values = values,
                 children = children
             };
 
+            ////TODO: come up with nice icons depicting different control types
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/Internal/InputControlTreeView.cs:199]: Was this TODO added as part of this work? Some associated task that should be filed? Feels unrelated?

**Review by ekcoh — APPROVED** [review]
Nice work! Would be very beneficial to have performance metrics in place for this kind of work to see effect of introduced changes see comment in review.

File: Packages/com.unity.inputsystem/InputSystem/Editor/Internal/InputControlTreeView.cs (hunk 20 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Editor/Internal/InputControlTreeView.cs:199]: It's unrelated, just moved code around

File: Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs (hunk 18 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs]: Can fix, but dunno, most game devs know that vec4 is 128 bit, kinda a muscle memory I feel

File: Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs (hunk 17 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs:79]: We have no idea why is that, and we have no idea if anyone is relying on this. So I can't add a comment

File: Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs (hunk 2 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs:79]: I will keep this one because we have practical experience from dots mathematics where inlining helped, but will remove ones on ReadValue

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 10 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:895]: Keep as is for now, probably safer that way as `ApplyParameterChanges` on root (device) will force-recompute all the states, so kinda an escape hatch of sorts

File: Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs (hunk 2 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/ButtonControl.cs:79]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 11 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:939]: Now it sets correct value `optimizedControlDataType` after printing the error.
Fixed

File: Packages/com.unity.inputsystem/CHANGELOG.md (hunk 13 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/CHANGELOG.md]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 15 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:923]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs (hunk 18 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs]: Fixed

File: Packages/com.unity.inputsystem/InputSystem/Controls/Vector3Control.cs (hunk 19 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/Vector3Control.cs]: Fixed

File: Assets/Tests/InputSystem/CoreTests_Controls.cs (hunk 12 above)
Comment (jimon) [reply Assets/Tests/InputSystem/CoreTests_Controls.cs:1421]: We haven't found any good existing tests for this.
But I've added two performance tests.

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs (hunk 14 above)
Comment (duckets) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:892]: Is optimizedControlDataType and CalculateOptimizedControlDataType() intended to be used by end-users of the package? Or is this what "function shouldn't be called manually" refers to? (manually is confusing)

From the changelog it sounds like end users should only need to use SetInternalFeatureFlag and ApplyParameterChanges to actually use this feature, is this correct?

Also does this new feature have any corresponding documentation in the manual?

Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:892]: Depends, if a user only uses built-in control types (`Vector2Control`, `AxisControl`, etc) then they only need `SetInternalFeatureFlag` and `ApplyParameterChanges` to get started.

If user wants to enable this functionality for their own control types, they need to provide `CalculateOptimizedControlDataType` implementation.

Comment (duckets) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:892]: I've updated the docs based on discussions with @jimon

**Review by duckets — APPROVED** [review]
Approved with docs amendments
